fluence-js/3_in_depth.md

```diff
@@ -350,3 +350,177 @@ Tetraplets have the form of:
+   * Signs the data using key pair's private key. Required by aqua
+   */
+  async sign(
+    data: number[],
+    callParams: CallParams<"data">
+  ): Promise<SignResult> {
+    // ...
+  }
+
+  /**
+   * Verifies the signature. Required by aqua
+   */
+  verify(signature: number[], data: number[]): Promise<boolean> {
+    // ...
+  }
+}
+```
+
+`securityGuard` specifies the way the `sign` method checks where the incoming data is allowed to be signed or not. It accepts one argument: call params (see "Call params and tetraplets") and must return either true or false. Any predicate can be specified. Also, FluenceJS is shipped with a set of useful predicates:
+
+```typescript
+/**
+ * Only allow calls when tetraplet for 'data' argument satisfies the predicate
+ */
+export const allowTetraplet = (pred: (tetraplet: SecurityTetraplet) => boolean): SigSecurityGuard => {/*...*/};
+
+/**
+ * Only allow data which comes from the specified serviceId and fnName
+ */
+export const allowServiceFn = (serviceId: string, fnName: string): SigSecurityGuard => {/*...*/};
+
+/**
+ * Only allow data originated from the specified json_path
+ */
+export const allowExactJsonPath = (jsonPath: string): SigSecurityGuard => {/*...*/};
+
+/**
+ * Only allow signing when particle is initiated at the specified peer
+ */
+export const allowOnlyParticleOriginatedAt = (peerId: PeerIdB58): SigSecurityGuard => {/*...*/};
+
+/**
+ * Only allow signing when all of the predicates are satisfied.
+ * Useful for predicates reuse
+ */
+export const and = (...predicates: SigSecurityGuard[]): SigSecurityGuard => {/*...*/};
+
+/**
+ * Only allow signing when any of the predicates are satisfied.
+ * Useful for predicates reuse
+ */
+export const or = (...predicates: SigSecurityGuard[]): SigSecurityGuard => {/*...*/};
+};
+```
+
+Predicates as well as the `Sig` definition can be found in `@fluencelabs/fluence/dist/services`
+
+`Sig` class is accompanied by `registerSig` which allows registering different signing services with different keys. The mechanism is exactly the same as with ordinary aqua services e.g:
+
+```typescript
+// create a key per from pk bytes
+const customKeyPair = await KeyPair.fromEd25519SK(pkBytes);
+
+// create a signing service with the specific key pair
+const customSig = new Sig(customKeyPair);
+
+// restrict sign usage to our needs
+customSig.securityGuard = allowServiceFn("my_service", "my_function");
+
+// register the service. Please note, that service id ("CustomSig" here) has to be specified.
+registerSig("CustomSig", customSig);
+```
+
+for a non-default peer, the instance has to be specified:
+
+```typescript
+const peer = new FluencePeer();
+await peer.start();
+
+// ...
+
+registerSig(peer, "CustomSig", customSig);
+```
+
+FluencePeer is shipped with the default signing service implementation, registered with id "Sig". Is is useful to work with Trust Graph and Registry API. The default service provides the following restrictions on the `sign` operation:
```

Comment:
FluencePeer is ships with the default signing service registered with id "Sig".
TrustGraph 
default service is ?